indexd/src/app_client.rs

```diff
@@ -43,63 +96,207 @@ pub struct SlabPinParams {
 }
 
 impl Client {
-    #[allow(dead_code)]
-    pub fn new(url: String, password: Option<String>) -> Self {
-        Self {
+    pub fn new<U: IntoUrl>(url: U, app_key: PrivateKey) -> Result<Self> {
+        Ok(Self {
             client: reqwest::Client::new(),
-            url: url.trim_end_matches('/').to_string(),
-            password,
+            url: url.into_url()?,
+            app_key,
+        })
+    }
+
+    /// Checks if the application is authenticated with the indexer. It returns
+    /// true if authenticated, false if not, and an error if the request fails.
+    pub async fn check_app_authenticated(&self, status_url: &Url) -> Result<bool> {
+        let resp = self
+            .client
+            .get(status_url.clone())
+            .query(&self.sign(
+                status_url.clone(),
+                Method::POST,
+                None,
+                OffsetDateTime::now_utc(),
+            ))
+            .send()
+            .await?;
+        match resp.status() {
+            StatusCode::UNAUTHORIZED => Ok(false),
+            StatusCode::NO_CONTENT => Ok(true),
+            _ => Err(Error::ApiError(resp.text().await?)),
         }
     }
 
-    #[allow(dead_code)]
+    /// Checks if an auth request has been approved. If the auth request is
+    /// still pending, it returns false.
+    pub async fn check_request_status(&self, status_url: &Url) -> Result<bool> {
```

Comment:
These urls should be owned so they can potential be moved instead of always needing to clone them before passing into reqwest.